### ISSUE TITLE
Generate SBOM and sign release artefacts

### ISSUE THREAD

**Issue by richardcase** [kubernetes-sigs/cluster-api-provider-aws#3325]
/kind feature
/area release
/area security
/help
/priority important-soon
/triage accepted

**Describe the solution you'd like**
We should be generating a SBOM for CAPA and also signing this and any other release artefacts.

**Anything else you would like to add:**
We should probably use [sigstore](https://docs.sigstore.dev/)

**Environment:**

- Cluster-api-provider-aws version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

**Comment by sedefsavas** [comment]
Do you have any example how other projects in Kubernetes ecosystem does this?

**Comment by richardcase** [comment]
We use it in FluxCD

**Comment by sedefsavas** [comment]
Relevant to this, there is an effort going on in K8s https://github.com/kubernetes/release/issues/2383

Looks like sigstore is being used there: https://github.com/kubernetes/website/pull/31610/files

**Comment by sedefsavas** [comment]
Might worth to come up with a common workflow for cluster-api and other providers too.

**Comment by richardcase** [comment]
> Might worth to come up with a common workflow for cluster-api and other providers too.

I agree @sedefsavas . We'll probably have to make changes to image-builder / the image promoter stuff which would touch all the providers (probably)

**Comment by sedefsavas** [comment]
There is a nice TGIK talk about what's being done in Kubernetes about this: https://www.youtube.com/watch?v=H1D0fk9sZ8I

**Comment by puerco** [comment]
Hey I just saw this issue referenced in SIG Release, I'm happy to help out!

**Comment by richardcase** [comment]
This is important for the future, so

/reopen

**Comment by flxw** [comment]
Hi @richardcase ,
I could help with this one. Would you be ok with using [sigstore](sigstore.dev)?
That would allow you to have publicly verifiable signatures without key management.
How do you create your releases? I quickly checked the workflows, but couldn't find any dedicated workflow for that.

I'd like to split the SBOM generation ticket into a separate issue because we can take care of that easily once Sigstore is in place.
What do you think?

**Comment by sbueringer** [comment]
I wonder if there is already infrastructure in place in the kubernetes community that we can just use?

e.g. I noticed that our container images are already signed because signing was added to the Kubernetes image promotion process.

**Comment by flxw** [comment]
I think you are referring to [this](https://blog.sigstore.dev/kubernetes-signals-massive-adoption-of-sigstore-for-protecting-open-source-ecosystem-73a6757da73), right? That's the same mechanism I would love to use :)

**Comment by sbueringer** [comment]
Ah perfect. Thx for the info, I"m not really familar with how it works :)

**Comment by flxw** [comment]
@sedefsavas - I saw that your name is on most of the releases. Could you kindly give me context on how those are authored? I couldn't find a Github Actions workflow that created the release, so I am assuming it's manual?
My idea is to add a step for SBOM generation and upload into the Sigstore infrastructure into the release process.
Looking forward to your answer!

**Comment by richardcase** [comment]
> @sedefsavas - I saw that your name is on most of the releases. Could you kindly give me context on how those are authored? I couldn't find a Github Actions workflow that created the release, so I am assuming it's manual? My idea is to add a step for SBOM generation and upload into the Sigstore infrastructure into the release process. Looking forward to your answer!

@flxw  - we follow these steps when doing a release: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/docs/book/src/development/releasing.md

So manual with some automation.

**Comment by flxw** [comment]
Hi! My apologies for the long lead time, but I finally got around to making some time for this.
This exemplary commit integrates `cosign` into the Makefile, adding a separate `release-signed` step.
While it's totally possible to execute this manually on a local machine, I would recommend implementing this as a Github Action. These have an OIDC provider that gives us provenance information for free. This information will be engraved into Sigstore during the signing process.

https://github.com/kubernetes-sigs/cluster-api-provider-aws/commit/c340edba366aa101f36e4109ce2477f8174e7b3b

However, I've also seen that there is a larger movement for Kubernetes artifact signing underway:
https://github.com/kubernetes/enhancements/issues/3031

I'll link up with the people on that issue, as I hope to solve this a bit more elegantly and with benefits for the other projects.
What do you think?

**Comment by richardcase** [comment]
Thanks for the update @flxw.

It would be good to be aligned with the wider Kubernetes community effort on artifact signing.

**Comment by furkatgofurov7** [comment]
Hey! 

We have a similar tracking [issue](https://github.com/kubernetes-sigs/cluster-api/issues/6153) in CAPI to have this in place, and +1 from my side to have a common workflow for it. But going through the discussion quickly, it seems to me only the signing part of the SBoM was discussed, however how about the SBoM generation itself? 
We had a quick chat earlier with @cpanato and got to know, the upstream k8s community uses https://github.com/kubernetes-sigs/bom to generate it, so perhaps that is the workflow we could follow for SBoM generation.

**Comment by richardcase** [comment]
@furkatgofurov7 - thanks for input and the link to the k8s community bom is really helpful.

**Comment by furkatgofurov7** [comment]
This is still relevant, so

/triage accepted


@flxw hi! 

> However, I've also seen that there is a larger movement for Kubernetes artifact signing underway: [kubernetes/enhancements#3031](https://github.com/kubernetes/enhancements/issues/3031)
> 
> I'll link up with the people on that issue, as I hope to solve this a bit more elegantly and with benefits for the other projects. What do you think?

Just wanted to check on the status of this work since this is needed not only in CAPA but in CAPI also. 
As I checked https://github.com/kubernetes/enhancements/issues/3031 is almost near finishing line (KEP + code + docs merged) and in the Beta status, but the only item open I see is https://github.com/kubernetes/release/issues/2286 where the discussion is ongoing (TL;DR sign SBOM after generation with bom tool using cosign). 

Would be great to hear your and others opinion on this one, and probably we can already start laying a foundation for the SBOM generation using the bom tool to start with and later improve it with cosign. Thoughts?

**Comment by richardcase** [comment]
@furkatgofurov7 - this would be good to discuss at the next CAPA office hours.

**Comment by furkatgofurov7** [comment]
> @furkatgofurov7 - this would be good to discuss at the next CAPA office hours.

@richardcase sure, I will try to make it to the upcoming office hours.
